src/charmed_kubeflow_chisme/testing/cos_integration.py

```diff
@@ -0,0 +1,226 @@
+import logging
+from pathlib import Path
+from typing import Any, Dict, Set
+
+import yaml
+from juju.action import Action
+from juju.application import Application
+from juju.model import Model
+from juju.unit import Unit
+
+log = logging.getLogger(__name__)
+
+GRAFANA_AGENT_APP = "grafana-agent-k8s"
+METRICS_ENDPOINT_RELATION = "metrics-endpoint"
+GRAFANA_DASHBOARD_RELATION = "grafana-dashboard"
+ALER_RULES_DIRECTORY = Path("./src/prometheus_alert_rules")
+
+
+async def deploy_and_assert_grafana_agent(
+    model: Model,
+    app: str,
+    channel: str = "latest/stable",
+    dashboard: bool = True,
+    metrics: bool = True,
+) -> None:
+    """Deploy grafana-agent-k8 and add relate it with app.
+
+    Helper function to deploy and relate grafana-agent-k8s with provided app.
+
+    Args:
+        model: libjuju model object
+        app: name of application with which the Grafana agent should be related
+        channel: grafana-agent-k8s channel name, latest/stable as default
+        dashboard: relate <app>:grafana-dashboard grafana-agent-k8s:grafana-dashboards-consumer
+        metrics: relate <app>:metrics-endpoint grafana-agent-k8s:metrics-endpoint
+    """
+    assert app in model.applications, f"application {app} was not found"
+
+    log.info("deploying %s from %s channel", GRAFANA_AGENT_APP, channel)
+    await model.deploy(GRAFANA_AGENT_APP, channel=channel)
+
+    if dashboard is True:
+        log.info(
+            "juju relate %s:%s %s:grafana-dashboards-consumer",
+            app,
+            GRAFANA_DASHBOARD_RELATION,
+            GRAFANA_AGENT_APP,
+        )
+        await model.add_relation(
+            f"{app}:{GRAFANA_DASHBOARD_RELATION}",
+            f"{GRAFANA_AGENT_APP}:grafana-dashboards-consumer",
+        )
+
+    if metrics is True:
+        log.info(
+            "juju relate %s:%s %s:metrics-endpoint",
+            app,
+            METRICS_ENDPOINT_RELATION,
+            GRAFANA_AGENT_APP,
+        )
+        await model.add_relation(
+            f"{app}:{METRICS_ENDPOINT_RELATION}", f"{GRAFANA_AGENT_APP}:metrics-endpoint"
+        )
+
+    await model.wait_for_idle(apps=[GRAFANA_AGENT_APP], status="blocked", timeout=5 * 60)
+
+
+async def _check_metrics_endpoint(app: Application, metrics_endpoint: str) -> None:
+    """Check metrics endpoint accessibility.
+
+    Checking accessibility of metrics endpoint from grafana-agent-k8s. If metrics endpoint is
+    defined as `*:5000/metrics` it will be changed to `<app-name>.<namespace>.svc:5000/metrics`.
+    """
+    if metrics_endpoint.startswith("*"):
+        url = f"http://{app.name}.{app.model.name}.svc{metrics_endpoint[1:]}"
+    else:
+        url = f"http://{metrics_endpoint}"
+
+    cmd = f"curl -m 5 -sS {url}"
+    grafana_agent_app = app.model.applications[GRAFANA_AGENT_APP]
+    log.info("testing metrics endpoint with cmd: `%s`", cmd)
+    for unit in grafana_agent_app.units:
+        await _run_on_unit(unit, cmd)
```

Comment:
I don't think that curling `http://k8s-address` will work since the `svc-name.namespace.svc` is included only in the k8s DNS. The workaround here (as we discussed) would be to curl through the a k8s pod (grafana-agent-k8s unit?).

Reply:
We are doing that, we are running on grafana-agent-k8s unit. I think by default it's on charm and not service container, but that should be enough, or not?

Reply:
You 're right. I probably didn't notice that this runs on the unit!